iOS_SDK/OneSignalSDK/OneSignalUser/Source/Requests/OSRequestCustomEvents.swift

```diff
@@ -0,0 +1,87 @@
+ to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ copies of the Software, and to permit persons to whom the Software is
+ furnished to do so, subject to the following conditions:
+
+ 1. The above copyright notice and this permission notice shall be included in
+ all copies or substantial portions of the Software.
+
+ 2. All copies of substantial portions of the Software may only be used in connection
+ with services provided by OneSignal.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+ */
+
+import OneSignalCore
+import OneSignalOSCore
+
+class OSRequestCustomEvents: OneSignalRequest, OSUserRequest {
+    var sentToClient = false
+    let stringDescription: String
+    override var description: String {
+        return stringDescription
+    }
+
+    var identityModel: OSIdentityModel
+
+    func prepareForExecution(newRecordsState: OSNewRecordsState) -> Bool {
+        if let onesignalId = identityModel.onesignalId,
+           newRecordsState.canAccess(onesignalId),
+           let appId = OneSignalConfigManager.getAppId()
+        {
+            _ = self.addPushSubscriptionIdToAdditionalHeaders()
+            self.path = "apps/\(appId)/integrations/custom_events"
+            return true
+        } else {
+            return false
+        }
+    }
+
+    init(events: [[String: Any]], identityModel: OSIdentityModel) {
+        self.identityModel = identityModel
+        self.stringDescription = "<OSRequestCustomEvents with events: \(events)>"
+        super.init()
+        self.parameters = [
+            "events": events
+        ]
+        self.method = POST
+    }
+
+    func encode(with coder: NSCoder) {
+        coder.encode(identityModel, forKey: "identityModel")
+        coder.encode(parameters, forKey: "parameters")
+        coder.encode(method.rawValue, forKey: "method") // Encodes as String
+        coder.encode(timestamp, forKey: "timestamp")
+    }
+
+    required init?(coder: NSCoder) {
+        guard
+            let identityModel = coder.decodeObject(forKey: "identityModel") as? OSIdentityModel,
+            let rawMethod = coder.decodeObject(forKey: "method") as? UInt32,
+            let parameters = coder.decodeObject(forKey: "parameters") as? [String: Any],
+            let timestamp = coder.decodeObject(forKey: "timestamp") as? Date
+        else {
+            // Log error
+            return nil
+        }
+        self.identityModel = identityModel
+        self.stringDescription = "<OSRequestCustomEvents with parameters: \(parameters)>"
+        super.init()
```

Comment:
nit, calling super.init() seems out of place in the middle of the method, I would recommend calling it at the top or bottom of method.

Reply:
Yeah it feel weird but it needs to be there in the middle. The stuff above it are properties on `this` class and the ones below are on the superclass. 
- If you put it at the top, you'll get an error like "Property 'self.stringDescription' not initialized at super.init call"
- If you put it at the bottom, it'll error with "'self' used in property access 'parameters' before 'super.init' call"